File: cpp/muha_slon.cpp
// Make sure to run this program from inside the "data" directory that contains
// dictionary.txt

#define MAX_WORDS 2048
#define START_WORD "муха"
#define END_WORD "слон"
#define WORD_LEN 4
#define WORD_BYTES WORD_LEN*2+1 // UTF8 2-byte characters

#include <stdio.h>
#include <string.h>
#include <assert.h>
#include <stdlib.h>

#define ArrayCount(Array) (sizeof(Array) / sizeof((Array)[0]))

int
StringLengthUTF8(char *String)
{
    int Result = 0;

    while(*String)
    {
        if(*String >= 0 && *String <= 127)
        {
            String += 1;
        }
        else if((*String & 0xE0) == 0xC0)
        {
            String += 2;
        }
        else if((*String & 0xF0) == 0xE0)
        {
            String += 3;
        }
        else if((*String & 0xF8) == 0xF0)
        {
            String += 4;
        }
        ++Result;
    }

    return(Result);
}

void
RemoveNewline(char *String)
{
    while(*String)
    {
        if(*String == '\n')
        {
            *String = '\0';
            break;
        }
        ++String;
    }
}

bool
DifferByOneLetter(char *Word1, char *Word2)
{
    int DifferenceCount = 0;

    while(*Word1 && *Word2)
    {
        // UTF8 2-byte characters
        if((*Word1 != *Word2) || (*(Word1+1) != *(Word2+1)))
        {
            ++DifferenceCount;
        }
        Word1 += 2;
        Word2 += 2;
    }

    bool Result = DifferenceCount == 1;

    return(Result);
}

struct word_node
{
    int WordIndex;
    word_node *Next;
};

int
main()
{
    char Words[MAX_WORDS][WORD_BYTES];
    int WordCount = 0;
    word_node *AdjacencyLists[MAX_WORDS];
    int EdgeTo[MAX_WORDS];
    bool Marked[MAX_WORDS];
    word_node *BFSQueueHead = 0;
    word_node *BFSQueueTail = 0;
    int StartWordIndex = -1;
    int EndWordIndex = -1;

    char Word[128];
    FILE *DictionaryFile = fopen("dictionary.txt", "r");

    if(!DictionaryFile)
    {
        printf("Cannot open dictionary.txt\n");
        return(1);
    }

    while(fgets(Word, ArrayCount(Word), DictionaryFile))
    {
        RemoveNewline(Word);
        if(StringLengthUTF8(Word) == WORD_LEN)
        {
            strcpy(Words[WordCount++], Word);
        }
    }

    fclose(DictionaryFile);

    for(int WordIndex = 0;
        WordIndex < WordCount;
        ++WordIndex)
    {
        for(int WordIndex2 = 0;
            WordIndex2 < WordCount;
            ++WordIndex2)
        {
            if (WordIndex != WordIndex2)
            {
                if(DifferByOneLetter(Words[WordIndex], Words[WordIndex2]))
                {
                    word_node *Node = (word_node *)malloc(sizeof(word_node));
                    Node->WordIndex = WordIndex2;

                    if(AdjacencyLists[WordIndex])
                    {
                        Node->Next = AdjacencyLists[WordIndex];
                        AdjacencyLists[WordIndex] = Node;
                    }
                    else
                    {
                        Node->Next = 0;
                        AdjacencyLists[WordIndex] = Node;
                    }
                }
            }
        }
    }

    for(int WordIndex = 0;
        WordIndex < WordCount;
        ++WordIndex)
    {
        if(strcmp(Words[WordIndex], START_WORD) == 0)
        {
            StartWordIndex = WordIndex;
            break;
        }
    }

    if(StartWordIndex < 0)
    {
        printf("Couldn't find start word.\n");
        return(1);
    }

    for(int WordIndex = 0;
        WordIndex < WordCount;
        ++WordIndex)
    {
        if(strcmp(Words[WordIndex], END_WORD) == 0)
        {
            EndWordIndex = WordIndex;
            break;
        }
    }

    if(EndWordIndex < 0)
    {
        printf("Couldn't find end word.\n");
        return(1);
    }

    Marked[StartWordIndex] = true;
    BFSQueueHead = (word_node *)malloc(sizeof(word_node));
    BFSQueueHead->WordIndex = StartWordIndex;
    BFSQueueHead->Next = 0;

    while(BFSQueueHead)
    {
        int WordIndex = BFSQueueHead->WordIndex;
        word_node *HeadNext = BFSQueueHead->Next;
        free(BFSQueueHead);
        BFSQueueHead = HeadNext;

        for(word_node *AdjacentNode = AdjacencyLists[WordIndex];
            AdjacentNode;
            AdjacentNode = AdjacentNode->Next)
        {
            if(!Marked[AdjacentNode->WordIndex])
            {
                Marked[AdjacentNode->WordIndex] = true;
                EdgeTo[AdjacentNode->WordIndex] = WordIndex;

                word_node *QueueNode = (word_node *)malloc(sizeof(word_node));
                QueueNode->WordIndex = AdjacentNode->WordIndex;
                QueueNode->Next = 0;

                if (!BFSQueueHead)
                {
                    BFSQueueHead = BFSQueueTail = QueueNode;
                }
                else
                {
                    BFSQueueTail->Next = QueueNode;
                    BFSQueueTail = QueueNode;
                }
            }
        }
    }

    if(Marked[EndWordIndex])
    {
        for(int WordIndex = EndWordIndex;
            ;
            WordIndex = EdgeTo[WordIndex])
        {
            printf("%s\n", Words[WordIndex]);
            if(WordIndex == StartWordIndex)
            {
                break;
            }
        }
    }
    else
    {
        printf("Can't convert start word to the end word.\n");
    }

    return(0);
}

File: php/muha_slon.php
<?php

define('START_WORD', 'муха');
define('END_WORD', 'слон');
define('WORD_LEN', 4);
define('WORD_LEN_IN_BYTES', WORD_LEN*2); // UTF-8 2-byte characters

function main()
{
    mb_internal_encoding("UTF-8");

    //
    // Put all 4-letter words from dictionary into array.
    //

    $words = array();
    $dictionaryFile = fopen(dirname(__FILE__) . '/../data/dictionary.txt', 'r');

    if (!$dictionaryFile) {
        print "Couldn't open dictionary file!\n";
        return;
    }

    while (($word = fgets($dictionaryFile)) !== false) {
        $word = trim($word);
        if (mb_strlen($word) == WORD_LEN) {
            $words[] = $word;
        }
    }

    fclose($dictionaryFile);

    $startWordIndex = array_search(START_WORD, $words);
    if (!$startWordIndex) {
        print "Couldn't find start word in the dictionary!\n";
        return;
    }

    $endWordIndex = array_search(END_WORD, $words);
    if (!$endWordIndex) {
        print "Couldn't find end word in the dictionary!\n";
        return;
    }

    $wordsCount = count($words);

    //
    // Generate a graph of words, adding edges between words that differ by
    // one letter.
    //

    $adjacencyLists = array();
    $marked = array();
    $edgeTo = array();

    for ($wordIndex1 = 0; $wordIndex1 < $wordsCount; ++$wordIndex1) {
        for ($wordIndex2 = 0; $wordIndex2 < $wordsCount; ++$wordIndex2) {
            if ($wordIndex1 == $wordIndex2) {
                continue;
            }

            $differentLettersCount = 0;
            // Using mb_strlen is costly here, so we compare raw bytes,
            // assuming 2-byte cyrillic characters.
            // This version is ~3 times faster than mb_strlen version (17s -> 6s).
            for ($letterIdx = 0; $letterIdx < WORD_LEN_IN_BYTES; $letterIdx += 2) {
                if (($words[$wordIndex1][$letterIdx] != $words[$wordIndex2][$letterIdx]) ||
                    ($words[$wordIndex1][$letterIdx+1] != $words[$wordIndex2][$letterIdx+1])) {
                    ++$differentLettersCount;
                    if ($differentLettersCount > 1) {
                        break;
                    }
                }
            }

            if ($differentLettersCount == 1) {
                $adjacencyLists[$wordIndex1][] = $wordIndex2;
            }
        }
    }

    //
    // Use breadth-first search to find the shortest path from the start word
    // to the end word.
    //

    $bfsQueue = array();
    $marked[$startWordIndex] = true;
    array_push($bfsQueue, $startWordIndex);

    while ($bfsQueue) {
        $wordIndex = array_shift($bfsQueue);
        foreach ($adjacencyLists[$wordIndex] as $adjacentIndex) {
            if (isset($marked[$adjacentIndex])) {
                continue;
            }

            $marked[$adjacentIndex] = true;
            $edgeTo[$adjacentIndex] = $wordIndex;
            array_push($bfsQueue, $adjacentIndex);
        }
    }

    //
    // Output the path if it exists.
    //

    if (!isset($marked[$endWordIndex])) {
        print "Can't convert start word to the end word.\n";
        return;
    }

    $path = array();
    $wordIndex = $endWordIndex;
    while (true) {
        $path[] = $words[$wordIndex];
        if ($wordIndex == $startWordIndex) {
            break;
        }
        $wordIndex = $edgeTo[$wordIndex];
    }

    while ($path) {
        printf("%s\n", array_pop($path));
    }
}

main();
